### ISSUE TITLE
some latex strings can not be correctly rendered

### ISSUE THREAD

**Issue by thudjx** [MakieOrg/Makie.jl#1991]
By now, I have found 2 latex characters that can not be correctly rendered. The MWE is
```
fig = Figure(resolution=(100,100))
Label(fig[1,1], L"\partial")
save("mwe.pdf", fig)
```
, which results in
![image](https://user-images.githubusercontent.com/32637321/170739552-9a5ce85e-fae4-4f1a-8422-1be6ede08f37.png)
And similarly, `L"\degree"` is rendered into
![image](https://user-images.githubusercontent.com/32637321/170740823-a879315b-9a18-4b65-9ac4-f10d3d5638bb.png)

The infomation of `julia` is
```
julia> versioninfo()
Julia Version 1.7.3
Commit 742b9abb4d (2022-05-06 12:58 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: AMD EPYC 7542 32-Core Processor
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, znver2)
```
and the versions of packages are:
```
[13f3f980] CairoMakie v0.8.3
[b964fa9f] LaTeXStrings v1.3.0
[0a4f8689] MathTeXEngine v0.4.0
```
I have tried to generate tex elements by:
```
julia> generate_tex_elements(L"\partial")
1-element Vector{Any}:
 (TeXChar '∂' [U+2202 in cmmi10 - Regular], [0.0, 0.0], 1.0)

julia> generate_tex_elements(L"\degree")
1-element Vector{Any}:
 (TeXChar '°' [U+00B0 in cmmi10 - Regular], [0.0, 0.0], 1.0)
```
, which are normal.
And to check if the `.ttf` files are correct:
```
with_theme(Theme(font=joinpath((dirname ∘ dirname)(pathof(Makie.MathTeXEngine)), "assets", "fonts", "ComputerModern", "cmmi10.ttf"))) do
           fig = Figure()
           ax = Axis(fig[1,1])
           fig
           end
```
, which also works fine.

**Comment by fatteneder** [comment]
Just tested and can confirm that this happens with `CairoMakie` but not with `GLMakie`.

**Comment by fatteneder** [comment]
Regarding `degree`: I wonder if this is even supported by `MathTeXEngine`, because according to https://en.wikipedia.org/wiki/Degree_symbol#Software-specific it is only provided by the `gensymb` package.
As a workaround you could use `^{\circ}`. I just tested and it renders in both `GLMakie` and `CairoMakie`

![image](https://user-images.githubusercontent.com/26469701/170784706-40f20915-6d9c-42df-b6b6-d841a70c0149.png)

**Comment by fatteneder** [comment]
Also: `\partial` renders correctly to `PDF` with
```julia
[13f3f980] CairoMakie v0.8.0
[b964fa9f] LaTeXStrings v1.3.0
[0a4f8689] MathTeXEngine v0.4.0
```
as can be seen here (second to last slide): https://github.com/fatteneder/MakieSlides.jl/blob/ff2615ed9678abd6f9bd55baac80265531678ddb/examples/presentation.pdf

**Comment by AshtonSBradley** [comment]
also https://github.com/JuliaPlots/Makie.jl/issues/1967 

evident in these examples

https://lazarusa.github.io/BeautifulMakie/ScattersLines/MultipleMulticoloredLineCbarAround/
https://lazarusa.github.io/BeautifulMakie/ScattersLines/latexSample/

**Comment by thudjx** [comment]
@fatteneder It can be checked that `L"\degree"` can be correctly generated:
```
julia> generate_tex_elements(L"\degree")
1-element Vector{Any}:
 (TeXChar '°' [U+00B0 in cmmi10 - Regular], [0.0, 0.0], 1.0)
```
although the result is different fromn:
```
julia> generate_tex_elements(L"^\circ")
1-element Vector{Any}:
 (TeXChar '∘' [U+00B1 in cmsy10 - Regular], [0.12000000178813934, 0.0], 0.6)
```

**Comment by jkrumbiegel** [comment]
I vaguely remember that @Kolaru said that some difficulty lies in the way needed symbols are accessed in the computer modern fonts. The pipeline we use accesses glyphs from FreeType fonts via unicode Chars. However, imagine a font that has two different glyphs for the character 'a', you would not be able to pick the correct glyph by passing the Char. You instead need an integer specifying the position of the glyph in the font. And I can imagine that there would be more symbols to access for MathTeXEngine if Makie allowed glyph indexing by integer instead of by Char. Buy that's just an educated guess from my side.

**Comment by Kolaru** [comment]
That's a very interesting one.

The pipeline for symbol finding in MathTexEngine is currently
1. Parse command e.g. `\alpha` or `α` (unicode input are supported)
2. In both case convert it to a common reprensentation (something like `(command=\alpha, char=α)`
3. Look at the ComputerModern char table if it contains the command
3.1. If yes, use the info there (specific font + glyph index in the font)
3.2. If no, use the char in the default font

The problem is just that `\degree` and `\partial` don't have the proper mapping in the CM char table. That should be an easy fix.

> However, imagine a font that has two different glyphs for the character 'a', you would not be able to pick the correct glyph by passing the Char. You instead need an integer specifying the position of the glyph in the font.

This problem is why we need step 3 above and we can not use a unicode font directly. It is a bit trickier than that even, since, as far as I understood, different glyphs for the same char are *at the same index* in the font (`Char` are already just integers pointing to the glyph number in a font, the proper char association being guaranteed for proper unicode fonts).

**Comment by fatteneder** [comment]
> However, imagine a font that has two different glyphs for the character 'a', you would not be able to pick the correct glyph by passing the Char. You instead need an integer specifying the position of the glyph in the font.

Isn't this why Unicode variation sequences exist? https://freetype.org/freetype2/docs/reference/ft2-glyph_variants.html
If I understand correctly then you could select the right glyph by providing a base char and then a selector taken from the range `U+E0100-U+E01EF`.
 
I am currently in the process of understanding how unicode emojis can be rendered, because some of those come with such a sequence (usually flags, peoples and their skin tones). However I am having trouble in using the `FreeType` API here.

**Comment by Kolaru** [comment]
> Isn't this why Unicode variation sequences exist? https://freetype.org/freetype2/docs/reference/ft2-glyph_variants.html
> If I understand correctly then you could select the right glyph by providing a base char and then a selector taken from the range `U+E0100-U+E01EF`.

Oh yes you are right. After some looking I also see that the variation glyphs seem to just be outside of the unicode indices in the font, so it may be possible to do it with Makie and Cairo actually.

Still, I haven't check if FreeTypeAbstraction.jl supports the variation sequence.

**Comment by fatteneder** [comment]
I think it does not, but there is a feature request from you :) https://github.com/JuliaGraphics/FreeTypeAbstraction.jl/issues/59

**Comment by jkrumbiegel** [comment]
One problem is that GLMakie will not be able to render emoji with the SDF pipeline.

**Comment by fatteneder** [comment]
Indeed, I was made aware of this at https://github.com/fatteneder/MakieSlides.jl/pull/15#issuecomment-1137872633

We are working around this for now using `scatter` and use actual `png`s of the emojis as markers.

The problem I had with `FreeType` was that I could not even query a glyph index for an unicode variation sequence. I still think it was due to me not really understanding `FreeType`, but I did not follow up on this due to the rendering issue you mention.
